runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java

```diff
@@ -135,6 +136,9 @@ public ExecutableStageDoFnOperator(
     this.contextFactory = contextFactory;
     this.outputMap = outputMap;
     this.sideInputIds = sideInputIds;
+
+    // Never chain because chaining could break concurrent element processing and state access
+    setChainingStrategy(ChainingStrategy.NEVER);
```

Comment:
This doesn't look quite right yet:

There are legitimate cases for chaining with different operator upstream of the executable stage (such as a native transform like the Kinesis consumer)

It is only a concern for executable stages that use stateful processing. If KeyedStateBackend is shared between operators in a chain, then we should set the chaining strategy only in such case.

Reply:
See https://github.com/apache/beam/pull/6726#discussion_r227817028